### ISSUE TITLE
Conflict with a CUDA-11 PyTorch installation

### ISSUE THREAD

**Issue by bryant1410** [dmlc/xgboost#10729]
XGBoost for Python depends on `nvidia-nccl-cu12`, which is for CUDA 12. I have a PyTorch 2.4.0 installation for CUDA 11.8, but when I use distributed mode, PyTorch picks up on the one installed by XGBoost for NCCL and it gives me problems for my environment.

My workaround is to install the CPU-only version of XGBoost. However, I still want to use XGBoost with CUDA support. It'd be nice if I could use it with `nvidia-nccl-cu11` instead. Not sure what the solution could be (maybe optional groups of dependencies for XGBoost, such as a `cu11` one, etc; or a different package). Note this could be a future problem when CUDA 13 comes out as well.

**Comment by trivialfis** [comment]
Thank you for raising an issue. I don't know what is the right solution either. As for a workaround, one can either use conda/mamba, or install xgboost without dependency (`pip install --no-deps`), then pickup these dependencies later.

The binary wheel itself can use `cu11`, it's just how to specify it the `pyproject.toml`.

**Comment by trivialfis** [comment]
May I ask what specific problem you ran into?

**Comment by bryant1410** [comment]
After doing a PyTorch distributed mode init, when doing a [`broadcast_object_list`](https://pytorch.org/docs/stable/distributed.html#torch.distributed.broadcast_object_list), I ran into some weird NCCL error:

```
  File "file.py", line 2, in function
    dist.broadcast_object_list(objects, src=src)
  File "env/lib/python3.10/site-packages/torch/distributed/c10d_logger.py", line 79, in wrapper
    return func(*args, **kwargs)
  File "env/lib/python3.10/site-packages/torch/distributed/distributed_c10d.py", line 2901, in broadcast_object_list
    broadcast(object_sizes_tensor, src=src, group=group)
  File "env/lib/python3.10/site-packages/torch/distributed/c10d_logger.py", line 79, in wrapper
    return func(*args, **kwargs)
  File "env/lib/python3.10/site-packages/torch/distributed/distributed_c10d.py", line 2205, in broadcast
    work = default_pg.broadcast([tensor], opts)
torch.distributed.DistBackendError: NCCL error in: ../torch/csrc/distributed/c10d/NCCLUtils.hpp:275, unhandled cuda error (run with NCCL_DEBUG=INFO for details), NCCL version 2.22.3
ncclUnhandledCudaError: Call to CUDA function failed.
Last error:
Cuda failure 'CUDA driver version is insufficient for CUDA runtime version'
```

It got fixed after removing `nvidia-nccl-cu12` installed by xgboost installation.

**Comment by trivialfis** [comment]
Closing as there's no solution in XGBoost. Feel free to reopen if there are suggestions.

**Comment by ikrommyd** [comment]
@trivialfis I was gonna open a separate issue for this, but why not make the `nvidia-nccl-cu12` installation optional and prompt the user to install it when it's required. I know there is a cpu only build but what if the user only has one GPU. They don't need `nccl` for that right? A) It's an extra 200MB wheel to download and install and B) it can cause conflicts (even though `xgboost` doesn't put any restriction on the version of `nvidia-nccl-cu12`) when a user wants to have something else installed with `CUDA 11.x` in the same environment.

**Comment by ikrommyd** [comment]
A solution to the second problem would be https://github.com/dmlc/xgboost/issues/10807 but that still leaves you with problem A) installing a large package they you may not need.

**Comment by trivialfis** [comment]
>  I was gonna open a separate issue for this, but why not make the nvidia-nccl-cu12 installation optional

We used to link NCCL as a static library into XGBoost. We moved to using the pip package in the hope that we can share the dependency with others and hence reduce the size of the user environment. Removing NCCL entirely would be a significant breaking change from that perspective.

I don't have an opinion on this for now. It's just that pypi is not a multi-language package index, for finer control, one can use other tools like conda.

**Comment by ikrommyd** [comment]
Well you don't always need NCCL  right? You can do most of the xgboost stuff without it no?

**Comment by trivialfis** [comment]
@hcho3 has more insight into the wheel next https://github.com/wheelnext , feel free to comment if there's anything actionable for XGBoost.

**Comment by hcho3** [comment]
@trivialfis Yes, the next standard for packaging should obviate the need for having to create multiple package names (`xgboost-cpu`, `xgboost-cu11`, `xgboost-cu12` etc). But it will be a while until the standard is finalized and ratified.

Currently, there are two alternatives:
* Create separate packages: `xgboost-cpu`, `xgboost-cu11`, `xgboost-cu12`, as proposed in #10807. OR
* Make NCCL an extra dependency, e.g. `xgboost[nccl-cu11]`, `xgboost[nccl-cu12]`. The disadvantage is that users who just install `xgboost` will not be able to use multiple GPUs right away.

**Comment by ikrommyd** [comment]
If I had to bet, I would bet that most users have a single GPU so I think the second option is more optimal and when a user tries to use multiple GPUs, a prompt would just tell them to install the nvidia nccl wheel.
